### PR TITLE
fix: retry cockroachDB serialization failures in generated API handlers

### PR DESCRIPTION
**Pull request by JuanJoseRestrepo**
### Closes #471

###Summary

Adds github.com/cockroachdb/cockroach-go/v2 v2.4.3 and emits every write in the generated API handlers inside crdbgorm.ExecuteTx. The change lives in the SDK handler generation, not in the generated files; all 18 handler files were regenerated from it.

After this change no generated handler writes outside a retry. There are 236 retry closures, and no remaining h.RequestDB(c) call ending in Create, Save, Updates or Delete.

- [x] 1. AddXxx     -> 52 closures, wrapping Create
- [x] 2. UpdateXxx  -> 52 closures, wrapping Updates
- [x] 3. ReplaceXxx -> 52 closures, wrapping Save
- [x] 4. DeleteXxx  -> 80 closures, wrapping Delete, plus the Updates that schedules a deletion for reconciled types

**What goes inside the closure, and why it differs**

The issue asks for the whole read-modify-write to be wrapped. That is right where the write consumes the read, and unnecessary where it does not, so the four shapes are not treated identically.

UpdateXxx wraps the read and the write. Updates(&updated) is applied to the row loaded by First, so a restart that re-ran only the write would apply it on top of a stale read.

AddXxx wraps the duplicate-name read and the create. Discussed under "Beyond the issue's scope" below.

DeleteXxx and ReplaceXxx wrap only the write. The reads that precede them do not feed them. The plain delete reads and then deletes by primary key; the reconciled delete reads and then writes constant scheduling values; replace reads and then saves with updatedXxx.ID = existingXxx.ID, where that key comes from the URL. Re-running the write alone is correct in all three. Including the reads would have meant merging deleteObjectChecks and deleteObjectExecution, which are built from two independent conditionals giving four combinations, and lifting the 404 and 409 responses out of the middle of the flow. Agreed with @lander2k2 that the uniformity was not worth that.

**The NATS publish stays outside**

In the reconciled delete path the publish sits immediately after the scheduling update, in the same block. Wrapping the block rather than the statement would have re-published the deletion notification on every retry and woken the controller more than once for a single deletion. A check across all 236 closures confirms none contains a JS.Publish.

**Two details the issue does not call out**

c.Bind cannot be retried. It drains the request body and fails with 400 EOF on a second attempt, so the payload check and the bind run once, before the transaction. readBody, which PayloadCheck uses, does restore the body, and PayloadCheck reaches the object only through reflect.TypeOf, so passing the zero value instead of the row read from the database is equivalent.

h.RequestDB(c) is not just the connection. It applies WithContext and QueryScopes(c). ExecuteTx opens the transaction itself, so h.DB is passed instead and the query scopes are applied to tx inside the closure.

**Beyond the issue's scope**

The AddXxx handler checks for a duplicate name before creating. That read was running outside any transaction, so besides not being retried it left a window where two concurrent creates could both find the name free. Read and create now share one transaction, which closes that race as well as absorbing the 40001.

**A bug introduced and fixed during testing**

The first version of the add handlers hoisted scopedDB := tx.Scopes(...) and reused it for the duplicate-name read and the create. gorm's Scopes does not open a new session, so the statement carries its error forward. On a create with a free name the read fails with ErrRecordNotFound by design, and that error stayed on scopedDB and skipped the create entirely: every POST answered 500 record not found.

Each operation now builds its own instance from tx. The update handlers had the same shape; it never surfaced there because their read returns early on error, but the fragility was identical and is gone too.

Unit tests could not have caught this, because they have no database. The integration suite against a live control plane failed deterministically on all three runs, and passes three for three after the fix.

**Behavior change**

Moving the payload check ahead of the read in UpdateXxx changes one error precedence: a malformed payload against an ID that does not exist now answers 400 where it used to answer 404. No test depends on that precedence, and 400 describes a malformed request more accurately, but it is a change.

**Verification**

Build clean. 16 unit test packages green. gofmt unchanged from the branch point. Codegen idempotent.

The integration suite was run against a freshly built control plane three times consecutively, all passing, with no SQLSTATE 40001. Before this change the same suite produced database serialization conflict in one of two runs, which is the failure that motivated the issue. Three clean runs is not proof that it cannot recur under heavier contention, and the reconciler's requeue-on-error backstop remains the last resort as the issue notes.

**Notes for review**

The retry policy is the library default: 50 attempts with no backoff between them. That is aggressive for sustained contention, since it holds the request open while hammering the database. crdb.WithMaxRetries(ctx, n) or WithRetryPolicy would set something deliberate. I left the default rather than pick a number without data, but it is worth a decision.

Two commit subjects are wrong and should be read against their contents rather than their titles: f2c49ab says "delete handlers" but contains the add and delete work, and 55d469e also says "delete handlers" but contains replace. The commit bodies were lost when they were written, so the reasoning above does not appear in the history.

**Not addressed**

The generated clients carry 260 unguarded response.Data[0] accesses. They are all GetByID, Create, Update, Replace and Delete, where the server either answers 404 or returns the object it acted on, so none is the filter-query shape that can answer 200 with an empty Data. Defensive hardening for the SDK template rather than a live bug, and unrelated to this issue.

### CODE REVIEW

**Review by lander2k2 — CHANGES_REQUESTED** [review]
There are custom hand-written handlers that weren't addressed.  These handlers sit alongside the SDK-generated API handlers in the same `pkg/api-server/v0/handlers/` directory.

These handlers need the same retries applied before we merge this PR.

File: pkg/api-server/v0/handlers/module.go [hunk 1]
```diff
@@ -43,17 +44,27 @@ func (h Handler) AddModuleApiRouteWithModuleObjectReferences(c echo.Context) err
...
+	// isolation CockroachDB answers a write conflict with SQLSTATE 40001 and
+	// expects the client to re-run it
+	if err := crdbgorm.ExecuteTx(
+		c.Request().Context(), h.DB, nil,
+		func(tx *gorm.DB) error {
+			// the database assigns the key, so a retried attempt must not carry
+			// the one a rolled-back attempt was given
+			moduleApiRoute.ID = nil
+			return tx.Omit("ModuleObjects.*").Create(&moduleApiRoute).Error
+		},
```
[the remark this answers is not in the document]
Comment (JuanJoseRestrepo) [reply pkg/api-server/v0/handlers/module.go:57]: Valid, and the hook does exactly that — but this predates the change. The previous code was a bare h.DB.Create, and gorm wraps a single Create in an implicit transaction, so AfterCreate already ran before commit and a failed commit already left the route installed. What the retry changes is the frequency: where one attempt used to fail and orphan the route, up to fifty now run and one of them usually commits, so the router ends up consistent more often than before. The leak only survives when all fifty exhaust, which is the same state the old code reached on the first failure.

Happy to close it here rather than leave it. The subtlety is that Path carries no unique index, so two routes may share one router key and a blanket RemoveRoute on failure would take out a sibling's live route. The correct restore is to capture whatever was registered for that path before the transaction and put it back on failure, which needs a small read accessor on ModuleRouter since routes is unexported. Say the word and I'll add it, or I can file it separately since it is not introduced here.

[the remark this answers is not in the document]
Comment (JuanJoseRestrepo) [reply pkg/api-server/v0/handlers/module.go:57]: Fixed in the latest commit: the handler now snapshots whatever the router held for the path before the transaction and restores it on failure, rather than removing the key, so a sibling route sharing the path survives. ModuleRouter.Route is new for that, and module_router_test.go covers both the introduced-route and shared-path cases.

File: pkg/api-server/v0/handlers/module.go [hunk 2]
```diff
@@ -43,17 +44,42 @@ func (h Handler) AddModuleApiRouteWithModuleObjectReferences(c echo.Context) err
...
+	); err != nil {
+		// put the router back the way the hook found it
+		if hadPriorRoute {
+			api_v0.ModRouter.AddRoute(*moduleApiRoute.Path, priorRoute)
+		} else {
+			api_v0.ModRouter.RemoveRoute(*moduleApiRoute.Path)
+		}
```
[the remark this answers is not in the document]
Comment (JuanJoseRestrepo) [reply pkg/api-server/v0/handlers/module.go]: Right, and this one is mine — the snapshot goes stale, so the restore can delete a concurrent create's committed route. That is worse than the leak it was meant to fix, so I have reverted it. Reconciling against the database instead is equally racy, since a concurrent commit can land between the select and the write, and the only airtight version holds a process-wide lock across a transaction that retries up to fifty times.

The root cause is that afterCreate mutates the process-wide router from inside the transaction, so any compensation from the handler is racing it. That predates this PR: the previous h.DB.Create ran the same hook inside gorm's implicit transaction, so a failed commit already orphaned the route. The retry work leaves it strictly no worse, and in practice better, since most attempts now commit where one used to fail outright. Filing the hook design separately rather than growing this PR.

**Comment by JuanJoseRestrepo** [comment]
> There are custom hand-written handlers that weren't addressed. These handlers sit alongside the SDK-generated API handlers in the same `pkg/api-server/v0/handlers/` directory.
> 
> These handlers need the same retries applied before we merge this PR.

The hand-written handlers are addressed. Three of the six in that directory write to the database — kubernetes_workload.go, module.go and secret.go — and all three now wrap their writes in crdbgorm.ExecuteTx like the generated ones. The other three don't touch it.

Two things came out of it. AddKubernetesWorkloadResourceDefinitions opened a transaction and then wrote through h.DB instead of tx, so its "all or none at all" comment never held; that's fixed. And secret.go needed its Data field restored per attempt, since BeforeCreate nils it and a retry would otherwise have created a secret with no data.

I also added an integration test that drives about ninety transaction restarts through one row and asserts none surfaces to the client. Raising its load past the library's fifty-attempt cap makes it fail, so it isn't a no-op — which also answers the retry-policy question: the default of fifty exhausts at roughly five hundred concurrent writes to a single row. Happy to set something deliberate with crdb.WithMaxRetries if you'd prefer.

One thing I did not fix here: the module router is mutated from inside the transaction by the afterCreate hook. I tried compensating in the handler and reverted it, because every compensation races a concurrent commit. Filed as #492 — it predates this PR.

**Review by lander2k2 — APPROVED** [review]
